File: src/fusion/timestamp.ts
// Familista — Global timestamp synchroniser (Phase D-IP)
// ─────────────────────────────────────────────────────────────────────────
// Heterogeneous sensors arrive with heterogeneous clocks:
//   - GPS quartz oscillator (PPM range)
//   - ECG analogue front-end (jitters with skin contact)
//   - Neuromorphic vision (microsecond-accurate ASIC clock)
//   - Biochemical patch (slow polling, batched)
//   - Match operator (human-entered, second-precision)
//
// The fusion protocol requires every packet to be projected onto ONE
// monotonic axis: GlobalTimestampMs (ms since unix epoch, server-aligned).
//
// We estimate a linear correction per DeviceSession:
//     globalMs ≈ deviceUs/1000 + offsetMs(t) + drift_ppm · (t - t0)
//
// `offsetMs` is bootstrapped from the handshake (sessionKey response time
// vs ts received) and refined on every packet using the receiver timestamp.

import type { DeviceTimestamp, GlobalTimestampMs } from './types';

/** Per-session synchronisation state. */
interface SessionClock {
  /** Offset from device-monotonic to server-wall, ms. */
  offsetMs:   number;
  /** Optional drift rate (ppm) — slope of (device_clock - server_clock). */
  driftPpm:   number;
  /** Last packet device microsecond timestamp (for drift estimation). */
  lastDeviceUs:  number;
  /** Last packet server reception time (ms). */
  lastServerMs:  GlobalTimestampMs;
  /** Number of packets used so far for the estimate. */
  samples:    number;
}

const clocks = new Map<string, SessionClock>();

/** Smoothing factor for offset EMA. 0.05 = slow, stable. */
const EMA_ALPHA = 0.05;

/**
 * Initialise the clock for a session. Called once when the handshake
 * receives the first batch from a device.
 */
export function bootstrapClock(deviceSessionId: string, firstPacket: { deviceUs: number; serverRxMs: GlobalTimestampMs }): void {
  // Initial offset: server_now_ms - device_now_ms.
  const offsetMs = firstPacket.serverRxMs - firstPacket.deviceUs / 1000;
  clocks.set(deviceSessionId, {
    offsetMs,
    driftPpm:     0,
    lastDeviceUs: firstPacket.deviceUs,
    lastServerMs: firstPacket.serverRxMs,
    samples:      1,
  });
}

/**
 * Update the clock based on a newly received packet.
 *
 * If a device drifts faster than the server clock by N ppm, our offset
 * needs to be lowered over time. The drift estimator is a simple linear
 * regression in incremental form to keep memory O(1) per session.
 */
export function updateClock(deviceSessionId: string, packet: { deviceUs: number; serverRxMs: GlobalTimestampMs }): void {
  const cur = clocks.get(deviceSessionId);
  if (!cur) return bootstrapClock(deviceSessionId, packet);

  const newOffsetMs = packet.serverRxMs - packet.deviceUs / 1000;

  // Exponential moving average → resists jitter, follows slow drift.
  cur.offsetMs = cur.offsetMs * (1 - EMA_ALPHA) + newOffsetMs * EMA_ALPHA;

  // Drift estimate (ppm) between consecutive samples.
  if (cur.lastDeviceUs && cur.lastServerMs) {
    const dDev = (packet.deviceUs - cur.lastDeviceUs) / 1000; // ms
    const dSrv = packet.serverRxMs - cur.lastServerMs;
    if (dDev > 1 && dSrv > 1) {
      const sampleDriftPpm = ((dDev - dSrv) / dSrv) * 1_000_000;
      // Clamp pathological values; legitimate drift is typically <50 ppm.
      const clamped = Math.max(-200, Math.min(200, sampleDriftPpm));
      cur.driftPpm = cur.driftPpm * (1 - EMA_ALPHA) + clamped * EMA_ALPHA;
    }
  }

  cur.lastDeviceUs = packet.deviceUs;
  cur.lastServerMs = packet.serverRxMs;
  cur.samples += 1;
}

/**
 * Project a device-local microsecond timestamp onto the GlobalTimestamp axis.
 * If the session has no clock yet, falls back to "now".
 */
export function toGlobalMs(deviceSessionId: string, deviceUs: number): GlobalTimestampMs {
  const cur = clocks.get(deviceSessionId);
  if (!cur) return Date.now();
  // Project: deviceUs/1000 + offset, then subtract drift correction.
  const tDevMs = deviceUs / 1000;
  // Drift is in ppm: 1 ppm = 1µs per second. The longer the session, the
  // more we have to subtract. Use last-known anchor.
  const dtMs   = tDevMs - cur.lastDeviceUs / 1000;
  const driftCorrection = (cur.driftPpm * dtMs) / 1_000_000;
  return tDevMs + cur.offsetMs - driftCorrection;
}

/** Read-only inspector for diagnostics endpoints. */
export function inspectClock(deviceSessionId: string): DeviceTimestamp & { samples: number } | null {
  const cur = clocks.get(deviceSessionId);
  if (!cur) return null;
  return {
    deviceUs:   cur.lastDeviceUs,
    serverRxMs: cur.lastServerMs,
    offsetMs:   cur.offsetMs,
    driftPpm:   cur.driftPpm,
    samples:    cur.samples,
  };
}

/** Reset all in-memory clocks. Called only by tests. */
export function _resetAllClocks(): void { clocks.clear(); }

File: python/familista_fusion/core/timestamp.py
"""
GlobalTimestampSynchronizer — Python reference implementation.

Mirrors src/fusion/timestamp.ts. Heterogeneous sensors arrive with
heterogeneous clocks (GPS quartz, ECG AFE, neuromorphic ASIC, biochem
patch, human operator). Every packet must be projected onto ONE monotonic
axis before the fusion engine touches it:

    globalMs(p) ≈ deviceUs(p)/1000 + offsetMs(session) - drift_ppm·Δt/1e6

The estimator is per-DeviceSession: one EMA-smoothed offset and one
drift-ppm slope. Memory: O(1) per session.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Optional

from .packets import GlobalTimestampMs


EMA_ALPHA = 0.05      # smoothing factor — slow & stable
DRIFT_PPM_CLAMP = 200 # legitimate quartz drift is rarely > 50 ppm


@dataclass
class _SessionClock:
    offset_ms: float
    drift_ppm: float = 0.0
    last_device_us: int = 0
    last_server_ms: GlobalTimestampMs = 0
    samples: int = 0


class GlobalTimestampSynchronizer:
    """
    One instance per process. Carries an in-memory map
    device_session_id → SessionClock. Thread-safety is the caller's
    responsibility — the fusion engine accesses it from a single asyncio
    loop in production.
    """

    def __init__(self) -> None:
        self._clocks: Dict[str, _SessionClock] = {}

    def bootstrap(self, device_session_id: str, device_us: int, server_rx_ms: GlobalTimestampMs) -> None:
        offset = server_rx_ms - device_us / 1000.0
        self._clocks[device_session_id] = _SessionClock(
            offset_ms=offset,
            last_device_us=device_us,
            last_server_ms=server_rx_ms,
            samples=1,
        )

    def update(self, device_session_id: str, device_us: int, server_rx_ms: GlobalTimestampMs) -> None:
        cur = self._clocks.get(device_session_id)
        if cur is None:
            return self.bootstrap(device_session_id, device_us, server_rx_ms)

        new_offset = server_rx_ms - device_us / 1000.0
        cur.offset_ms = cur.offset_ms * (1 - EMA_ALPHA) + new_offset * EMA_ALPHA

        d_dev = (device_us - cur.last_device_us) / 1000.0
        d_srv = server_rx_ms - cur.last_server_ms
        if d_dev > 1 and d_srv > 1:
            sample_drift_ppm = ((d_dev - d_srv) / d_srv) * 1_000_000
            clamped = max(-DRIFT_PPM_CLAMP, min(DRIFT_PPM_CLAMP, sample_drift_ppm))
            cur.drift_ppm = cur.drift_ppm * (1 - EMA_ALPHA) + clamped * EMA_ALPHA

        cur.last_device_us = device_us
        cur.last_server_ms = server_rx_ms
        cur.samples += 1

    def to_global_ms(self, device_session_id: str, device_us: int) -> GlobalTimestampMs:
        cur = self._clocks.get(device_session_id)
        if cur is None:
            # Without an anchor, the best we can do is project as if offset is 0.
            return device_us // 1000
        t_dev_ms = device_us / 1000.0
        dt_ms    = t_dev_ms - cur.last_device_us / 1000.0
        drift_correction = (cur.drift_ppm * dt_ms) / 1_000_000.0
        return int(round(t_dev_ms + cur.offset_ms - drift_correction))

    def inspect(self, device_session_id: str) -> Optional[Dict[str, float]]:
        cur = self._clocks.get(device_session_id)
        if cur is None:
            return None
        return {
            "offset_ms":   cur.offset_ms,
            "drift_ppm":   cur.drift_ppm,
            "samples":     cur.samples,
            "last_device_us": cur.last_device_us,
            "last_server_ms": cur.last_server_ms,
        }

    def reset(self) -> None:
        """Test-only hook."""
        self._clocks.clear()
